Do sequential installs on editable dependencies

techalchemy opened the pull request:
- Fixes #1111 
- Uses existing infrastructure to add editable dependencies to the list
of sequentially installed packages
- Only alters functions used for this purpose
- Essentially there is a race condition because multiple editors are
trying to write to lib/site-packages/easy-install.pth for local editable
packages, so they overwrite one another

erinxocon commented:
@techalchemy Looks like we've got failing tests from this change!

techalchemy commented:
Yah forgot to change the function name in the test

techalchemy commented:
This appears permanently queued in appveyor.  It passes on my appveyor however:

[![Build status](https://ci.appveyor.com/api/projects/status/a383f9obssqlucox/branch/bugfix/1111-sequential-editable-installs?svg=true)](https://ci.appveyor.com/project/techalchemy/pipenv/branch/bugfix/1111-sequential-editable-installs)

gcarothers commented:
I can confirm this fixes #1111 on the original Pipfile and using a simple `pipenv install` as expected `easy-install.pth` contains all 15 entries.

nateprewitt commented:
Thanks for checking in on that @gcarothers!

I wrote up a response for this the other day and apparently didn't post it. To briefly summarize my thoughts, I think this solves the issue but from an architectural perspective, I'd really like to avoid this overloading what this function does. Our recent trend of repurposing functions to do multiple things, not because they're conceptually related, but because they happen to occur at the same place in the code is creating a lot of unnecessary coupling.

Something slightly cleaner might be to create something that does a general division/split operation and then implement precedence for how things are divided with ordered function calls. That'll allow us to scale a little easier in the future if we need more of these.

gcarothers commented:
I mean architecturally the fact that all locally referenced packages still have to go in the same easy-install.pth seems like an artifact of a bad old time ;) In the meantime without this fix it's not really possible for any project with N local packages to use pipenv?

nateprewitt commented:
@gcarthors it's definitely possible, you just need to pass --sequential with your install. We're just fixing bad defaults here causing a race condition with local package heavy installs.

techalchemy commented:
@nateprewitt I refactored just this one thing for the moment, for readability, and split functionality.  There are other places where we can use this type of refactoring but one thing at a time

techalchemy commented on pipenv/utils.py:
```diff
@@ -917,21 +917,100 @@ def proper_case(package_name):
+        entries = input_file.get(section, {})
+        for k in list(entries.keys()):
+            if test_function(entries.get(k)):
+                split_dict[k] = entries.pop(k)
```
It doesn't, although my initial pass at it used `for k, v in entries.items()` which did cause a runtime error.  This mimics the prior pattern of iterating over the list of keys rather than the dictionary itself, which (should?) allow us to manipulate it as we iterate.  That's how we handled it in the past and it _was_ working ok, tests were passing for me, but that doesn't prove it won't break things.

nateprewitt replied:
You're right, copying it into the new list does resolve the issue.

nateprewitt commented on pipenv/utils.py:
```diff
@@ -950,11 +950,12 @@ def split_section(input_file, section_suffix, test_function):
-    # return the original file if we can't find any pipfile or lockfile sections
     else:
-        return input_file
+        if any(section in input_file for section in lockfile_sections):
```
Why did this need to be moved?

techalchemy replied:
Hm I think I misunderstood what you were asking for in your review where you asked me to move this into the else statement for readability. To clarify -- did you mean that you prefer to test something like `if not any(section in input_file for section in pipfile_sections+lockfile_sections):` and do the alternative checks afterward?

nateprewitt replied:
Sorry, I was simply asking that you move the comment into the else statement rather than having it out of alignment. There should need be any code changes.

techalchemy replied:
kk sorry, I thought the request was odd but it seemed simple enough so i just did (what i thought was) it
